### PR TITLE
Feature/new httperrors handling

### PR DESCRIPTION
**Pull request by slashformotion**

### CODE REVIEW

File: services/httperrors/httperrors_test.go [hunk 1]
```diff
@@ -3,7 +3,7 @@ package httperrors_test
 import (
```
Comment (Zorin95670) [inline services/httperrors/httperrors_test.go:3]: For all tests after some research please use
https://pkg.go.dev/github.com/stretchr/testify/assert
to make assertion in you unit test, it's more readable

File: services/httperrors/httperrors_test.go [hunk 2]
```diff
@@ -44,3 +51,24 @@ func TestWrite(t *testing.T) {
 		t.Error("the body should contains the jsonified HTTPError")
 	}
 }
+
+func TestNewErrorNotFound(t *testing.T) {
```
Comment (Zorin95670) [inline services/httperrors/httperrors_test.go:55]: TestNewErrorNotFound, TestNewInternalServerError, TestNewUnauthorizedError not relevant test, in this case we want to check if default value is well set

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: services/httperrors/httperrors.go [hunk 3]
```diff
@@ -3,32 +3,92 @@ package httperrors
...
 		Message:     message,
 		GolangError: golangError,
+		toLog:       toLog,
 	}
 }
 
 // Convert an HTTPError to a json string
-func (httpError *HTTPError) ToJSON() string {
+func (httpError *HTTPErrorImpl) ToJSON() string {
 	return fmt.Sprintf(`{"error": %q, "msg":%q, "status": %q}`, httpError.Err, httpError.Message, http.StatusText(httpError.Status))
```
[the remark this answers is not in the document]
Comment (slashformotion) [reply services/httperrors/httperrors.go]: https://pkg.go.dev/fmt
```
%q	a single-quoted character literal safely escaped with Go syntax.


```

File: services/httperrors/httperrors_test.go [hunk 4]
```diff
@@ -3,36 +3,62 @@ package httperrors_test
...
-
 func TestTojson(t *testing.T) {
-	error := httperrors.NewHTTPError(http.StatusBadRequest, "Error while parsing json", "The request body was malformed", nil)
+	err := "Error while parsing json"
+	message := "The request body was malformed"
+	error := httperrors.NewHTTPError(http.StatusBadRequest, err, message, nil, true)
 	if error.ToJSON() == "" {
 		t.Errorf("the json string returned by the ToJSON method should not return an empty string")
 	}
 	if !json.Valid([]byte(error.ToJSON())) {
```
Comment (Zorin95670) [inline services/httperrors/httperrors_test.go:24]: replace by `assert.True`

File: services/httperrors/httperrors_test.go [hunk 5]
```diff
@@ -3,36 +3,62 @@ package httperrors_test
...
-		t.Errorf("the HTTPError returned by the contructor should not be nil")
-	}
-}
-
 func TestTojson(t *testing.T) {
-	error := httperrors.NewHTTPError(http.StatusBadRequest, "Error while parsing json", "The request body was malformed", nil)
+	err := "Error while parsing json"
+	message := "The request body was malformed"
+	error := httperrors.NewHTTPError(http.StatusBadRequest, err, message, nil, true)
 	if error.ToJSON() == "" {
```
Comment (Zorin95670) [inline services/httperrors/httperrors_test.go:21]: replace this by `assert.NotEmpty`

File: services/httperrors/httperrors_test.go [hunk 6]
```diff
@@ -3,36 +3,62 @@ package httperrors_test
...
+
+}
+func TestError(t *testing.T) {
+	error := httperrors.NewHTTPError(http.StatusBadRequest, "Error while parsing json", "The request body was malformed", nil, true)
+	assert.Contains(t, error.Error(), error.ToJSON())
 }
 
 func TestWrite(t *testing.T) {
 	res := httptest.NewRecorder()
-	error := httperrors.NewHTTPError(http.StatusBadRequest, "Error while parsing json", "The request body was malformed", nil)
```
Comment (Zorin95670) [inline services/httperrors/httperrors_test.go:33]: use assert in this method

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: services/httperrors/httperrors_test.go (hunk 5 above)
Comment (slashformotion) [reply services/httperrors/httperrors_test.go]: Done

File: commands/rootCmd.go [hunk 7]
```diff
@@ -91,4 +77,30 @@ func init() {
 	rootCfg.GKey("database.sslmode", verdeter.IsStr, "", "[DB] the sslmode to use when connecting to the database server")
 	rootCfg.SetRequired("database.sslmode")
 
+	loggerModeKey := "logger.mode"
```
Comment (Zorin95670) [inline commands/rootCmd.go:80]: split this in different files: 
commands/database.go: func initDatabaseCommands(**) { // init all command of database}

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: router/middlewares/middlewareLogger.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"net/http"
+
+	"github.com/ditrit/badaas/configuration"
+	"github.com/noirbizarre/gonja"
+	"github.com/noirbizarre/gonja/exec"
+	"go.uber.org/zap"
+)
+
+func buildRequest(template *exec.Template, r *http.Request) string {
```
Comment (Zorin95670) [inline router/middlewares/middlewareLogger.go:12]: missing documentation, bad naming function, it's better to have `getLogMessage` or something like that

File: router/middlewares/middlewareLogger.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+	result, _ := template.Execute(
+		gonja.Context{
+			"proto":  r.Proto,
+			"method": r.Method,
+			"url":    r.URL.Path,
+		})
+	return result
+}
+
+func CreateLoggerMiddleware() func(next http.Handler) http.Handler {
```
Comment (Zorin95670) [inline router/middlewares/middlewareLogger.go:22]: missing documentation

File: router/middlewares/middlewareJson.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/ditrit/badaas/services/httperrors"
+)
+
+var HErrorCantMarshall httperrors.HTTPError = httperrors.NewInternalServerError(
+	"json marshall error",
+	"Can't marshall the object returned by the JSON handler",
+	nil,
+)
+
+// Marshall
```
Comment (Zorin95670) [inline router/middlewares/middlewareJson.go:16]: improve documentation

File: controllers/info.go [hunk 11]
```diff
@@ -1,26 +1,19 @@
...
-	"encoding/json"
 	"net/http"
 
-	"github.com/ditrit/badaas/persistence/models"
+	"github.com/ditrit/badaas/persistence/models/dto"
 	"github.com/ditrit/badaas/resources"
+	"github.com/ditrit/badaas/services/httperrors"
 )
 
 // Info controller, return json with status and version of api.
```
Comment (Zorin95670) [inline controllers/info.go:11]: need to change documentation

File: commands/initLoggerCommands.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+				}
+			}
+			return fmt.Errorf("logger.mode should be one of the implemented modes (%v)", allowedModes)
+
+		},
+	}
+	cfg.SetValidator(loggerModeKey, loggerModeValidator)
+
+	loggerRequestTemplateKey := "logger.request.template"
+	cfg.GKey(loggerRequestTemplateKey, verdeter.IsStr, "", "[LOGGER] the request logger template")
```
Comment (Zorin95670) [inline commands/initLoggerCommands.go:40]: `[LOGGER] the request logger template` -> `Template message for all request logs`

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: commands/initServerCommands.go [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package commands
+
+import (
+	"github.com/ditrit/verdeter"
+	"github.com/ditrit/verdeter/validators"
+)
+
+func initServerCommands(cfg *verdeter.VerdeterCommand) {
+	cfg.GKey("server.max_timeout", verdeter.IsInt, "", "maximum timeout (in second)")
```
Comment (Zorin95670) [inline commands/initServerCommands.go:9]: change by `server.timeout`, be more explicit on the description and add default in description

File: commands/initLoggerCommands.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	"fmt"
+
+	"github.com/ditrit/verdeter"
+	verdetermodels "github.com/ditrit/verdeter/models"
+)
+
+func initLoggerCommands(cfg *verdeter.VerdeterCommand) {
+
+	loggerModeKey := "logger.mode"
+	cfg.GKey(loggerModeKey, verdeter.IsStr, "", "[LOGGER] either \"prod\" or \"dev\"")
```
Comment (Zorin95670) [inline commands/initLoggerCommands.go:13]: rework on description to be like the others, add default in.

File: commands/initDatabaseCommands.go [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package commands
+
+import "github.com/ditrit/verdeter"
+
+func initDatabaseCommands(cfg *verdeter.VerdeterCommand) {
+	cfg.GKey("database.port", verdeter.IsInt, "", "[DB] the port of the database server")
```
Comment (Zorin95670) [inline commands/initDatabaseCommands.go:6]: remove all `[DB]`

File: commands/initServerCommands.go [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+	cfg.SetDefault("max_timeout", 15)
```
Comment (Zorin95670) [inline commands/initServerCommands.go:10]: for all this command, miss: `server.` when you set default

File: commands/initLoggerCommands.go [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"fmt"
+
+	"github.com/ditrit/verdeter"
+	verdetermodels "github.com/ditrit/verdeter/models"
+)
+
+func initLoggerCommands(cfg *verdeter.VerdeterCommand) {
+
+	loggerModeKey := "logger.mode"
```
Comment (Zorin95670) [inline commands/initLoggerCommands.go:12]: Why do you have a variable for the key and not for the other config ?

Comment (slashformotion) [reply commands/initLoggerCommands.go:12]: Now every config key use a variable

**Review by Zorin95670 — APPROVED** [review]

**Pull request merged**